example/socks/http.cpp

```diff
@@ -0,0 +1,355 @@
+
+        // Skip path.
+        std::advance(it, tmp_it - it);
+
+        // Skip '?'
+        it++;
+    }
+    else
+    {
+        tmp_it = std::find(it, url.end(), '#');
+        if (tmp_it != url.end())
+        {
+            // Read path.
+            for (auto first = it; first != tmp_it; first++)
+                result.path.push_back(*first);
+
+            // Skip path.
+            std::advance(it, tmp_it - it);
+
+            // Skip '#'
+            it++;
+
+            // Read fragment.
+            for (auto first = it; first != url.end(); first++)
+                result.fragment.push_back(*first);
+
+            return true;
+        }
+        else
+        {
+            // Read path.
+            for (auto first = it; first != url.end(); first++)
+                result.path.push_back(*first);
+
+            return true;
+        }
+    }
+
+    // Read query.
+    tmp_it = std::find(it, url.end(), '#');
+    if (tmp_it != url.end())
+    {
+        // Read query.
+        for (auto first = it; first != tmp_it; first++)
+            result.query.push_back(*first);
+
+        // Skip query.
+        std::advance(it, tmp_it - it);
+
+        // Skip '#'
+        it++;
+
+        // Read fragment.
+        for (auto first = it; first != url.end(); first++)
+            result.fragment.push_back(*first);
+
+        return true;
+    }
+
+    // Read query.
+    for (auto first = it; first != url.end(); first++)
+        result.query.push_back(*first);
+
+    return true;
+}
+
+//------------------------------------------------------------------------------
+
+// Report a failure
+void
+fail(boost::system::error_code ec, char const* what)
+{
+    std::cerr << what << ": " << ec.message() << "\n";
+}
+
+// Performs an HTTP GET and prints the response
+void
+do_session(
+    std::string const& host,
+    std::string const& port,
+    std::string const& target,
+    int version,
+    std::string const& socks_server,
+    boost::asio::io_context& ioc,
+    boost::asio::yield_context yield)
+{
+    boost::system::error_code ec;
+
+    // These objects perform our I/O
+    tcp::resolver resolver{ioc};
+    tcp::socket socket{ioc};
+
+    socks_url url;
+    if (!parse_socks_url(socks_server, url))
+        return fail(ec, "parse_socks_url");
+
+    // socks handshake.
+    int socks_version = url.scheme == "socks4" ? 4 : 5;
+    socks_version = url.scheme == "socks5" ? 5 : 0;
+    if (socks_version == 0)
+        return fail(ec, "incorrect socks version");
+
+    // Look up the domain name
+    auto const results = resolver.async_resolve(url.host, url.port, yield[ec]);
+    if (ec)
+        return fail(ec, "resolve");
+
+    // Make the connection on the IP address we get from a lookup
+    boost::asio::async_connect(socket, results.begin(), results.end(), yield[ec]);
+    if (ec)
+        return fail(ec, "connect");
+
+    socks::async_handshake(socket, host, static_cast<unsigned short>(std::atoi(port.c_str())),
```

Comment:
Where does one get a socks server to test with?

Reply:
If I am reading this right, then to operate a socks client you simply connect to the server and perform the handshake. From that point on, the socket behaves as a transparent tunnel? So no additional state has to be maintained for the lifetime of the connection, and the proxy connection algorithm can be expressed simply as a free function?